### DISCUSSION TITLE
Limit local cache size (cache eviction)

### DISCUSSION THREAD

**Discussion by mehulkar** [vercel/turborepo#7450] [Ideas]
### Discussed in https://github.com/vercel/turbo/discussions/1264

<div type='discussions-op-text'>

<sup>Originally posted by **maxburs** May 18, 2022</sup>
I'm caching the /node_modules/.turbo folder in CI. This is causing problems because of its size. I've seen it get as large at 4.5gb. Is there a good way to limit the size of the cache?</div>

<sub>[TURBO-2071](https://linear.app/vercel/issue/TURBO-2071/limit-local-cache-size-cache-eviction)</sub>

**Comment by anthonyshew** [comment]
We just spoke on this amongst the team and we've settled on that this is something that we should do. We have it on our board and we will get it prioritized at some point in the future. We're not sure when at the moment but having controls around cache eviction is something we're interested in being a feature.

**Comment by EloB** [comment]
@anthonyshew An idea would be to track when you actually hit a cache in like a json file with a key map of cache key and unix timestamp in the `.cache/turbo` folder like `hits.json`. So you can present how many unused/rarely used cache files you have at the moment.

Than maybe have the possibility to clean up unused files with a command for instance `turbo cache clean --unused` or something similar.

```
turbo build

...

 Tasks:    A successful, B total
Cached:    C cached, D total
Unused:    E caches since F ago, remove by "turbo cache clean --unused"
  Time:    Gs >>> FULL TURBO
```

**Comment by mehulkar** [comment]
That's a good idea and we want a mapping of `[code ref] -> [turbo hash]` for other reasons also.

**Comment by MisterJimson** [comment]
Found this thread after hitting 25GB. Would love this type of feature.

**Comment by vltansky** [comment]
@anthonyshew Is this feature still planned?

**Comment by tomzs** [comment]
Was wondering why computer ran out of space, turns out .turbo/cache got to 85GB. Would love this addition.

**Comment by alexwilson1** [comment]
Any update on this please? @anthonyshew

**Comment by Christophevvv** [comment]
+1
Also stumbled on this with computer running out of space .turbo/cache grew to 860 GiB. Woops

**Comment by anthonyshew** [comment]
Hey, folks, I'm going to make this a part of 2.10.

My only holdup at this point is I don't know what removal heuristic people would be happy with.

- Some arbitrary size limit?
- Expiration time?
- Max count of caches on disk?
- Something else?

Any of these could be configurable. Would love everyone's input.

**Reply by andrew-y0u** [reply]
In one of our projects, we keep the N most recent cache entries and delete the rest. Deleting entries not accessed in N days would also be useful.

**Reply by tomzs** [reply]
Similar approach here. We just delete entries that are older than week.

**Reply by mstuercke** [reply]
Size limit should be the most important one.
All other would be nice on top. It would be great to be able to configure multiple of these cases. If one of it matches, the cache entry should be deleted

**Reply by anthonyshew** [reply]
> Size limit should be the most important one.

What would you expect to happen if you try to write one cache that's bigger than the limit?

Describing the situation more verbose for clarity:
- You have zero caches. You've set your limit to 1GB.
- You run `turbo run build` for the first time ever.
- `turbo` attempts to write a cache that is 1.1GB.
- ??? <-- What goes here?

**Reply by mstuercke** [reply]
> 
> - ??? <-- What goes here?
> 

No cache will be written as it exceeds the limit you've set by yourself

**Reply by anthonyshew** [reply]
I'm conflicted. A lot of the reason to use Turborepo is for task caching. This would make it so you could never cache the task.

I suppose a warning would suffice. It is something set by user configuration, after all, and it's always within your power/choice to change your limit.

**Reply by mstuercke** [reply]
If you set the retention time to 0 days or the max caches to 0, you could not cache as well. 
If you set the cache to a max. of 1GB (e.g. because it's the limit of your filesystem), you don't want to cache, as you explicitly defined it like that. 

The default should be "unlimited" caching, as it is right now. I don't see an issue here

**Reply by anthonyshew** [reply]
Shipped! https://turborepo.dev/docs/reference/configuration#cachemaxage
